rpc/send.go

```diff
@@ -0,0 +1,108 @@
+package rpc
+
+import (
+	"context"
+	"unsafe"
+
+	"capnproto.org/go/capnp/v3/internal/mpsc"
+	rpccp "capnproto.org/go/capnp/v3/std/capnp/rpc"
+)
+
+type preparer interface {
+	Prepare(rpccp.Message) error
+}
+
+type sendQueue mpsc.Queue
+
+func newSendQueue() *sendQueue { return (*sendQueue)(mpsc.New()) }
+
+func (sq *sendQueue) Send(ctx context.Context, p preparer) (err error) {
+	cherr := cherrPool.Get()
+	defer cherrPool.Put(cherr)
+
+	sq.SendAsync(ctx, p, cherr)
+
+	select {
+	case err = <-cherr:
+	case <-ctx.Done():
+		err = ctx.Err()
+	}
+
+	return
+}
+
+func (sq *sendQueue) SendAsync(ctx context.Context, p preparer, r ErrorReporter) {
+	(*mpsc.Queue)(sq).Send(sendReq{
+		er:       r,
+		preparer: p,
+	})
+}
+
+func (sq *sendQueue) Recv(ctx context.Context) preparer {
+	v, err := (*mpsc.Queue)(sq).Recv(ctx)
+	if err != nil {
+		return sendFailure(err)
+	}
+
+	return *(*preparer)(unsafe.Pointer(&v))
+}
+
+type prepFunc func(rpccp.Message) error
+
+func (prepare prepFunc) Prepare(m rpccp.Message) error { return prepare(m) }
+
+func sendFailure(err error) prepFunc {
+	return func(rpccp.Message) error { return err }
```

Comment:
There is a lot of indirection in this file that as far as I can tell only serves to simplify the body of `Conn.send`. But my feeling is it would probably be better to have some if/elses in there than to do it this way; this seems very convoluted.